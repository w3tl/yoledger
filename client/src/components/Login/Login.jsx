import React from 'react';
import PropTypes from 'prop-types';
import { TextInput, Button } from '../Elements';

class Login extends React.Component {
  static propTypes = {
    signin: PropTypes.func.isRequired,
  }

  state = {
    username: '',
    password: '',
  }

  handleChange = name => ({ target }) => {
    this.setState({
      [name]: target.value,
    });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const { signin } = this.props;
    const { username, password } = this.state;
    signin({ username, password });
  }

  render() {
    const { username, password } = this.state;
    return (
      <form onSubmit={this.handleSubmit}>
        <TextInput name="username" label="Username" value={username} required type="text" onChange={this.handleChange('username')} />
        <TextInput name="password" label="Password" value={password} required type="password" onChange={this.handleChange('password')} />
        <Button type="submit">Signin</Button>
      </form>
    );
  }
}

export default Login;
