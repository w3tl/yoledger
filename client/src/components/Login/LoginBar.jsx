import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Menu } from 'semantic-ui-react';

function ProfileHeader({ signout }) {
  return (
    <Fragment>
      <Menu.Item name="profile" as={Link} to="/profile" />
      <Menu.Item name="signout" onClick={signout} />
    </Fragment>
  );
}

ProfileHeader.propTypes = {
  signout: PropTypes.func.isRequired,
};

function LoginBar({ signout, loggedIn }) {
  return (
    <Menu.Menu position="right">
      {loggedIn && <ProfileHeader signout={signout} />}
      {!loggedIn && (
        <Menu.Item as={Link} to="/signin" name="Sign-in" />)}
    </Menu.Menu>
  );
}

LoginBar.propTypes = {
  signout: PropTypes.func.isRequired,
  loggedIn: PropTypes.bool.isRequired,
};

export default LoginBar;
