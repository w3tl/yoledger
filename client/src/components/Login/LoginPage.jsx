import React from 'react';
import PropTypes from 'prop-types';
import { Button } from 'semantic-ui-react';
import { Form } from '../Elements';

const initialState = { username: '', password: '' };

class LoginPage extends React.Component {
  static propTypes = {
    signin: PropTypes.func.isRequired,
    loading: PropTypes.bool,
    error: PropTypes.object, // eslint-disable-line
    initialState: PropTypes.shape({
      username: PropTypes.string,
      password: PropTypes.string,
    }),
  }

  static defaultProps = { loading: false, initialState }

  // eslint-disable-next-line react/destructuring-assignment
  state = { ...initialState, ...this.props.initialState }

  handleChange = (e, { name, value }) => {
    this.setState({ [name]: value });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const { signin } = this.props;
    const { username, password } = this.state;
    signin({ variables: { login: username, password } });
  }

  render() {
    // TODO: check if loggedin
    const { username, password } = this.state;
    const isValid = username.length > 0 && password.length > 0;
    return (
      <Form onSubmit={this.handleSubmit} {...this.props}>
        <Form.Input
          label="Username"
          name="username"
          value={username}
          onChange={this.handleChange}
        />
        <Form.Input
          label="Password"
          name="password"
          value={password}
          type="password"
          onChange={this.handleChange}
        />
        <Button disabled={!isValid} type="submit">Sign in</Button>
      </Form>
    );
  }
}

export default LoginPage;
