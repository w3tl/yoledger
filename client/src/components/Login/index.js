import LoginBarComponent from './LoginBar';
import LoginPageComponent from './LoginPage';
import loginBarWithData from './LoginBarHOC';
import loginPageWithData from './LoginPageHOC';

const LoginBar = loginBarWithData(LoginBarComponent);
const LoginPage = loginPageWithData(LoginPageComponent);

export {
  LoginBar,
  LoginPage,
};
