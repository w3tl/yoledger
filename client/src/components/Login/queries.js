import gql from 'graphql-tag';

export const SIGNIN_MUTATION = gql`
mutation signin($login: String!, $password: String!) {
  signin(login: $login, password: $password) {
    token
  }
}
`;

export const SIGNOUT_MUTATION = gql`
mutation signout($token: String) {
  signout(token: $token) {
    success
  }
}
`;
