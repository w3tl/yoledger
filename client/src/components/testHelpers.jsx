import React from 'react';
import { shallow, mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

export const withProvider = Component => props => (
  <MockedProvider {...props}>
    <Component />
  </MockedProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return fixedDate;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
