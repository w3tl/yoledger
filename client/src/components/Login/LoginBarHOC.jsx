/* eslint-disable react/destructuring-assignment,react/prop-types */
import React from 'react';
import { Mutation } from 'react-apollo';
import { SIGNIN_MUTATION, SIGNOUT_MUTATION } from './queries';

const withSigninMutation = Wrapped => props => (
  <Mutation
    mutation={SIGNIN_MUTATION}
    onCompleted={(data) => {
      props.handleSignin(data.signin.token);
    }}
  >
    {(signin, { loading, error }) => {
      if (loading) return 'Sign in...';
      if (error) return error.message;
      return (
        <Wrapped
          signin={({ username, password }) => signin({
            variables: { login: username, password },
          })}
          {...props}
        />);
    }}
  </Mutation>
);

const withSignoutMutation = Wrapped => props => (
  <Mutation
    mutation={SIGNOUT_MUTATION}
    onCompleted={(data) => {
      if (data.signout.success) {
        props.handleSignout();
      }
    }}
  >
    {(signout, {
      loading, error,
    }) => {
      if (loading) return 'Sign out...';
      if (error) return error.message;
      return (
        <Wrapped
          signout={signout}
          {...props}
        />);
    }}
  </Mutation>
);

export default Component => withSigninMutation(withSignoutMutation(Component));
