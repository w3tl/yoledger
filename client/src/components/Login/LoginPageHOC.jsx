import React from 'react';
import { Mutation } from 'react-apollo';
import { SIGNIN_MUTATION } from './queries';

const withSigninMutation = Wrapped => ({ signin, ...props }) => (
  <Mutation
    mutation={SIGNIN_MUTATION}
    onCompleted={({ signin: { token } }) => signin(token)}
  >
    {(mutate, { loading, error }) => (
      <Wrapped signin={mutate} loading={loading} error={error} {...props} />
    )}
  </Mutation>
);

export default Component => withSigninMutation(Component);
