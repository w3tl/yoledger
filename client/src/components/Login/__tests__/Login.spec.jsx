import React from 'react';
import { mount } from 'enzyme';
import Login from '../Login';
import { testRenderWithoutError } from '../../testHelpers';

describe('Login', () => {
  testRenderWithoutError(<Login signin={() => {}} />);

  it('should handle submit', () => {
    const signin = jest.fn();
    const output = mount(
      <Login signin={signin} />,
    );
    output.find('TextInput[name="username"]').simulate('change', { target: { value: 'admin' } });
    output.find('TextInput[name="password"]').simulate('change', { target: { value: 'password' } });
    output.find('form').simulate('submit');
    expect(signin).toHaveBeenCalledWith({
      username: 'admin',
      password: 'password',
    });
  });
});
