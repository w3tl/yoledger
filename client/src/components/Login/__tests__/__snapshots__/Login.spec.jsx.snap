// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders without error 1`] = `
<form
  onSubmit={[Function]}
>
  <Component
    label="Username"
    name="username"
    onChange={[Function]}
    required={true}
    type="text"
    value=""
  />
  <Component
    label="Password"
    name="password"
    onChange={[Function]}
    required={true}
    type="password"
    value=""
  />
  <Button
    type="submit"
  >
    Signin
  </Button>
</form>
`;
